src/crowdin_client.cpp

```diff
@@ -135,150 +152,329 @@ CrowdinClient::~CrowdinClient() {}
+                pathStr += '/';
+                pathStr += path.top();
+                path.pop();
+            }
+            if(!pathStr.empty())
+                i.pathName = str::to_wstring(pathStr) + i.pathName;
+        }
+
+        //TODO: get more until all branches gotten (if more than 500)    
+        return m_api->get(url + "/branches?limit=500");
+    }).then([this, url, prj](json r) {
+        // Handle branches
+        std::map<int, std::string> branches;
+
+        for(const auto& i : r["data"])
+        {
+            const json& d = i["data"];
+            branches[d["id"]] = d["name"].get<std::string>();
+        }
+
+        for(auto& i : prj->files)
+            if(i.branchId != NO_ID)
+                i.pathName = "/" + str::to_wstring(branches[i.branchId]) + i.pathName;
+
+        return *prj;
+    });
 }
 
 
-dispatch::future<void> CrowdinClient::DownloadFile(const std::string& project_id,
-                                                   const std::wstring& file,
-                                                   const Language& lang,
+dispatch::future<void> CrowdinClient::DownloadFile(const long project_id,
+                                                   const long file_id,
+                                                   const std::wstring& file_name,
+                                                   const std::string& lang_tag,
                                                    const std::wstring& output_file)
 {
-    // NB: "export_translated_only" means the translation is not filled with the source string
-    //     if there's no translation, i.e. what Poedit wants.
-    auto url = "/api/project/" + project_id + "/export-file"
-                   "?json="
-                   "&export_translated_only=1"
-                   "&language=" + lang.LanguageTag() +
-                   "&file=" + http_client::url_encode(file);
-    return m_api->download(url, output_file);
+     wxLogTrace("poedit.crowdin", "Getting file URL: projects/%ld/translations/builds/files/%ld", project_id, file_id);
+    return RefreshToken().then([=] () {
+    return m_api->post(
+        "projects/" + std::to_string(project_id) + "/translations/builds/files/" + std::to_string(file_id),
+        json_data({
+            { "targetLanguageId", lang_tag },
+            // for XLIFF files should be exported "as is" so set to `false`
+            { "exportAsXliff", !wxString(file_name).Lower().EndsWith(".xliff.xliff") }
+        }))
+        .then([this, output_file] (json r) {
+            std::string url = r["data"]["url"],
+                   proto(wxString(url).BeforeFirst(':').mb_str()),
+                   host(wxString(url).AfterFirst('/').AfterFirst('/').BeforeFirst('/').mb_str());
+
+            auto downloader = std::make_shared<crowdin_http_client>(*this, proto + "://" + host);
+            wxLogTrace("poedit.crowdin", "Gotten file URL: %s", r.dump().c_str());
+            return downloader->download(url, output_file)
```

Comment:
I don't understand what's going on. Why do you explode the URL only to compose it back together on literally the next line?

In any case, if this *works* in real-life testing (particularly on Linux and Windows where C++ REST SDK is used; I know the underlying `http_client` on macOS is OK in this regard), i.e. passing *absolute URL* to `download()` works (my reading the library was that it won't), then there would be two possible simpler fixes:

1. Just ignore the `http_client` prefix URL mess, pass full URL to its constructor, then the same full URL to `download()`
2. Pass empty "" base URL to the ctor, full URL to `download()`
3. Pass full URL to the ctor, do `download("", …)`

Reply:
Actually, I forgot the most important one: *if* this works, then it would work on the API HTTP client and wouldn't require a transient one like this in the first place.

Reply:
Changed to `wxURI` in 199f93736a857f37a7a01438e9290a87558ca027 , also my answers below

> I don't understand what's going on. Why do you explode the URL only to compose it back together on literally the next line?

May be it will be more clear for you with `wxURI` since it was really hard to read as it was before, even for me after some time :), it was just my hurry to get everything just to work, sorry. 
If you also about `proto + "://" + host` here - none other way worked on all 3 platforms. While really tried exactly all 3 fixes proposed by you above but at least one of them did not work at least on one platform. Just lacked time to document behaviour which exactly on which platforms failed and which on which worked. But I really payed maximal attention to try to avoid URL exploding. Also I remember exactly that at least one of variants behaved differently even on Linux and Windows despite both use [cpprestsdk](https://github.com/microsoft/cpprestsdk) so far I realized what was final call for me to use as implemented with URL exploding (since started to think who know how some of backend might change behaviour in future while current current usage of client with `<proto>://domain` in constructor and URL in `download()` is most common expectations for most of clients lib even among different programming languages so must be future-proof for any backend.